pkg/apis/provisioning/v1alpha5/container_runtime_configuration.go

```diff
@@ -0,0 +1,61 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha5
+
+import (
+	"strings"
+
+	"knative.dev/pkg/apis"
+)
+
+// ContainerRuntimeConfiguration defines args to be used when configuring the Container Runtime.
+// Note, not all Container Runtime implementations will support all of these settings.
```

Comment:
Where is this struct from? Is this a generic interface?

Reply:
I have not found any existing API (struct) for those parameters. However, most Container Runtimes (CRI implementations) accept some configuration (usually via config files). This was inspired by [bottlerocket](https://github.com/bottlerocket-os/bottlerocket#container-image-registry-settings) which will then create the corresponding config files:
- For the Docker Container Runtime it will add it to [daemon-json](https://github.com/bottlerocket-os/bottlerocket/blob/v1.6.0/packages/docker-engine/daemon-json)
- For containerd it will be written to the [containerd-config](https://github.com/bottlerocket-os/bottlerocket/blob/v1.6.0/packages/containerd/containerd-config-toml_k8s#L34)

As Container Runtime configuration is not cloud provider specific, I thought that it would make sense to place it alongside `KubeletConfiguration` in the `Constraints` struct.

Reply:
I'm not comfortable defining this abstraction in Karpenter without deep understanding of where this is going. I worry that this will become an unmaintainable bag of parameters. 

Are you blocked on this? What's the use case.

Reply:
Yes, I am blocked on this. We need to be able to define the registry mirrors, to make sure all container are pulled through our own (pull through caching & security policy enforcing) registries.

We can also move it into AWS provider specific part (the `AWS` struct). However, as written before, this kind of settings are quite common among Container Runtime.

Reply:
Is registry mirrors all you needed? There's a lot that's being added here and I want to make sure we only add fields as they're necessary and prevent a bloated spec. 

Keep in mind that eventually we do intend on inlining UserData in the Provisioner as well which could be used to support any kind of customization (OS level or kubelet / container-runtime level).

Reply:
> Keep in mind that eventually we do intend on inlining UserData in the Provisioner as well which could be used to support any kind of customization (OS level or kubelet / container-runtime level).

@suket22 With this PR I just followed the path already paved (in the current code base), which does not support setting a verbatim UserData field and instead uses input from `KubeletConfiguration` and `AWS` structs to assemble the UserData. If you intend to replace all that logic with a simple `userData` field (e.g. in `Constraints` or `AWS` structs) that's fine and I am with you (and I will gladly refactor things with this PR). Just let me know which directions you guys want to go. 

> Is registry mirrors all you needed? There's a lot that's being added here and I want to make sure we only add fields as they're necessary and prevent a bloated spec.

While I understand your concerns, you should keep in mind that if you abstract over a configurable system (like for example AML2, bottlerocket, the kubelet, containerd, ...), there is a reason why those system have so many configuration parameters and that your abstractions must have means to pass on such configuration parameters to the underlying systems. Actually, this is a strong hint that [Approach 3 - Provide a more generic way to specify launch template contents in the Provider](https://github.com/aws/karpenter/pull/1270/files#diff-e7aba79d5b20d96be33150a7ca84bd17f4a5d0f52436b7498f540cee9b83ce28R110) might be the right way to avoid having to change karpenter with each user (specific) need. 

In our specific case we need to fine tune the kubelet and containerd to make sure we are compliant with several different high availability and security compliance requirements. For example we must be able to set `eventRecordQPS` (to unlimited) for security compliance reasons (see [CIS Kubernetes Benchmarks](https://www.cisecurity.org/benchmark/kubernetes) control 4.2.9).